### PR TITLE
Initial work on Constrained types

### LINKED ISSUE #10458: Constrained types support
Some types semantically have very limited set of possible values. Storing such kind of types can be beneficial for radare2 types and variables analyis. For example like "int < 250" or "char* [< 100]" where first is a type of integer that always less than 250, and a second - string which length is always less than 100 characters.
Especially those can be useful for initial analysis, when some library functions (or system calls) are known to return values limited somehow. Will make further autoanalysis easier.

### PR DESCRIPTION
**Pull request by sivaramaaa**
screenshot from 2018-08-19 13-21-38

Initial Work on Constrained types, still lot of test cases yet to handle (will be done in subsequent PR), i have added a few tests in r2r [#1442](https://github.com/radare/radare2-regressions/pull/1442) , which contains test for single and multiple conditions

This is handled in command `afta` , and controlled by config var `e anal.types.constraint`

Closes #10458

### CODE REVIEW

File: libr/core/anal_tp.c [hunk 1]
```diff
@@ -155,6 +155,80 @@ static ut64 get_addr(Sdb *trace, const char *regname, int idx) {
...
+	}
+	return res;
+}
+
+#define RKEY(a,k,d) sdb_fmt ("var.range.0x%"PFMT64x ".%c.%d", a, k, d)
+#define ADB a->sdb_fcns
+
+static void var_add_range (RAnal *a, RAnalVar *var, int cond, ut64 val) {
+	const char *key = RKEY (var->addr, var->kind, var->delta);
+	sdb_array_append_num(ADB, key, cond, 0);
```
Comment (radare) [inline libr/core/anal_tp.c]: Missing space before (

File: libr/core/anal_tp.c [hunk 2]
```diff
@@ -155,6 +155,80 @@ static ut64 get_addr(Sdb *trace, const char *regname, int idx) {
...
+	sdb_array_append_num(ADB, key, val, 0);
+}
+
+R_API char *var_get_constraint (RAnal *a, RAnalVar *var) {
+	char res[256] = {0};
+	const char *key = RKEY (var->addr, var->kind, var->delta);
+	int i, n = sdb_array_length(ADB, key);
```
Comment (radare) [inline libr/core/anal_tp.c]: Missing space before (

File: libr/core/anal_tp.c [hunk 3]
```diff
@@ -155,6 +155,80 @@ static ut64 get_addr(Sdb *trace, const char *regname, int idx) {
...
+	if (n < 2) {
+		return NULL;
+	}
+	for (i = 0; i < n; i += 2) {
+		ut64 cond = sdb_array_get_num (ADB, key, i, 0);
+		ut64 val = sdb_array_get_num (ADB, key, i + 1, 0);
+		switch (cond) {
+		case R_ANAL_COND_LE:
+			if (high) {
+				strncat (res, " && ", len);
```
Comment (radare) [inline libr/core/anal_tp.c]: Strncat is evil dont use it

File: libr/core/anal_tp.c [hunk 4]
```diff
@@ -155,6 +155,80 @@ static ut64 get_addr(Sdb *trace, const char *regname, int idx) {
...
+#define ADB a->sdb_fcns
+
+static void var_add_range (RAnal *a, RAnalVar *var, int cond, ut64 val) {
+	const char *key = RKEY (var->addr, var->kind, var->delta);
+	sdb_array_append_num(ADB, key, cond, 0);
+	sdb_array_append_num(ADB, key, val, 0);
+}
+
+R_API char *var_get_constraint (RAnal *a, RAnalVar *var) {
+	char res[256] = {0};
```
Comment (radare) [inline libr/core/anal_tp.c]: Dont use fixed size buffers

File: libr/core/anal_tp.c [hunk 5]
```diff
@@ -155,6 +155,80 @@ static ut64 get_addr(Sdb *trace, const char *regname, int idx) {
...
+	const char *key = RKEY (var->addr, var->kind, var->delta);
+	int i, n = sdb_array_length(ADB, key);
+	int len = sizeof (res) - 1;
```
Comment (radare) [inline libr/core/anal_tp.c]: This is const

File: libr/core/anal_tp.c [hunk 6]
```diff
@@ -155,6 +155,80 @@ static ut64 get_addr(Sdb *trace, const char *regname, int idx) {
...
+				strncat (res, " && ", len);
+			}
+			strncat (res, sdb_fmt ("<= 0x%"PFMT64x "", val), len);
+			low = true;
+			break;
+		case R_ANAL_COND_LT:
+			if (high) {
+				strncat (res, " && ", len);
+			}
+			strncat (res, sdb_fmt ("< 0x%"PFMT64x "", val), len);
```
Comment (radare) [inline libr/core/anal_tp.c]: Dont use strncat. Better use strbuf if u have to append many strings. Or rstring

File: libr/core/anal_tp.c [hunk 7]
```diff
@@ -155,6 +155,80 @@ static ut64 get_addr(Sdb *trace, const char *regname, int idx) {
...
+			high = true;
+			break;
+		}
+		if (low && high && i != n-2) {
+			strncat (res, " || ", len);
+			low = false;
+			high = false;
+		}
+	}
+	return strdup (res);
```
Comment (radare) [inline libr/core/anal_tp.c]: If u use strbuf you can drain it directly so there’s 0 copy involved in this step

File: libr/core/anal_tp.c [hunk 8]
```diff
@@ -155,6 +155,80 @@ static ut64 get_addr(Sdb *trace, const char *regname, int idx) {
...
+}
+
 static RList *parse_format(RCore *core, char *fmt) {
 	RList *ret = r_list_new();
```
Comment (radare) [inline libr/core/anal_tp.c]: Missing space before (

**Comment by radare** [comment]
i think appveyor is red because of their fault

**Comment by radare** [comment]
i wouldnt enable this in travis yet. wait to be merged and make another PR enabling it

**Comment by sivaramaaa** [comment]
> i wouldnt enable this in travis yet. wait to be merged and make another PR enabling it

yup , i agree on this , can u guys check this PR now !

**Pull request merged**